docs/scos/dev/keeping-a-project-upgradable/upgradability-guidelines/dependency-provider-contains-addition-logic.md

```diff
@@ -0,0 +1,73 @@
+
+Not all possible expressions are needed inside of the if statements for plugin registration and not all of them are supported. This check will verify that if IF construct is used for plugin registration, then only one of the following expressions is used:
+
+```php
+class_exists(\Foo\Bar::class) function call - it is required for BC reasons
+```
+    
+```php
+$this->getConfig()->isDevelopmentConsoleCommandsEnabled() function calls - it is required for plugins that are needed in development mode only (e.g.: profiling, debug, etc.)
+```
+
+## Example of code that causes an upgradability error:
+
+The method `getFormPlugins` in `FormDependencyProvider` contains unsupported expressions in IF construct `self::IS_DEV`.
+
+```php
+use Spryker\Yves\Form\FormDependencyProvider as SprykerFormDependencyProvider;
+
+class FormDependencyProvider extends SprykerFormDependencyProvider
+{
+    public const IS_DEV = true;
+    
+    ...
+    
+    protected function getFormPlugins(): array
+    {
+        $plugins = [
+            new ValidatorExtensionFormPlugin(),
+        ];
+        
+        $plugins[] = new CsrfFormPlugin();
+        
+        if (self::IS_DEV) {
+            $plugins[] = new WebProfilerFormPlugin();
+        }
+        
+        return $plugins;
+    }
+}
+```
+
+### Related error in the Evaluator output:
+
+```bash
+======================
+MULTIDIMENSIONAL ARRAY
+======================
+
++---+----------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------+
+| # | Message                                                                                                                    | Target                                                                                      |
++---+----------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------+
+| 1 | Reached max level of nesting for the plugin registration in the {FormDependencyProvider::getPlugins()}.                    | Pyz\Yves\Form\FormDependencyProvider\FormDependencyProvider        |
+|   | The maximum allowed nesting level is 2. Please, refactor code, otherwise it will cause upgradability issues in the future. |                                                                                             |
++---+----------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------+
+
+```
+
+### Resolving the error:
+
+To resolve the error provided in the example, try the following in the provided order:
+1. Try to have simple configuration arrays they shouldn't have more than 2 levels.
```

Comment:
`1. Try to have simple configuration arrays they shouldn't have more than 2 levels. => 
`1 - `Try to have a simple configuration of arrays. They shouldn't have more than 2 nested levels..`

Reply:
++